File: tests/Unit/IO/H5/Test_TensorData.py
# Distributed under the MIT License.
# See LICENSE.txt for details.

import unittest

import numpy as np
import numpy.testing as npt
from spectre.DataStructures import DataVector
from spectre.Domain import ElementId
from spectre.IO.H5 import ElementVolumeData, TensorComponent
from spectre.Spectral import Basis, Mesh, Quadrature


class TestTensorData(unittest.TestCase):
    # Tests for TensorComponent functions
    def test_tensor_component(self):
        # Set up Tensor Component
        tensor_component = TensorComponent("tensor component",
                                           DataVector([1.5, 1.1]))
        # Test name
        self.assertEqual(tensor_component.name, "tensor component")
        tensor_component.name = "new tensor component"
        self.assertEqual(tensor_component.name, "new tensor component")
        # Test data
        npt.assert_array_almost_equal(np.array(tensor_component.data),
                                      np.array([1.5, 1.1]))
        tensor_component.data = DataVector([6.7, 3.2])
        npt.assert_array_almost_equal(np.array(tensor_component.data),
                                      np.array([6.7, 3.2]))
        # Test str, repr
        self.assertEqual(str(tensor_component),
                         "(new tensor component, (6.7,3.2))")
        self.assertEqual(repr(tensor_component),
                         "(new tensor component, (6.7,3.2))")

    def test_element_volume_data(self):
        # Set up Extents and Tensor Volume data
        tensor_component_1 = TensorComponent("tensor component one",
                                             DataVector([1.5, 1.1]))
        tensor_component_2 = TensorComponent("tensor component two",
                                             DataVector([7.1, 5]))
        basis = Basis.Legendre
        quad = Quadrature.Gauss
        element_data = ElementVolumeData(
            element_name="grid_name",
            components=[tensor_component_1, tensor_component_2],
            extents=[3, 4],
            basis=2 * [basis],
            quadrature=2 * [quad])
        element_data_2 = ElementVolumeData(
            ElementId[2]("[B0,(L1I0,L0I0)]"),
            components=[tensor_component_1, tensor_component_2],
            mesh=Mesh[2]([3, 4], basis, quad))

        # Test extents
        self.assertEqual(element_data.extents, [3, 4])
        self.assertEqual(element_data_2.extents, [3, 4])
        element_data.extents = [5, 6]
        self.assertEqual(element_data.extents, [5, 6])
        # Test tensor components
        self.assertEqual(element_data.tensor_components,
                         [tensor_component_1, tensor_component_2])
        self.assertEqual(element_data_2.tensor_components,
                         [tensor_component_1, tensor_component_2])
        element_data.tensor_components = [
            tensor_component_1, tensor_component_1
        ]
        self.assertEqual(element_data.tensor_components,
                         [tensor_component_1, tensor_component_1])
        # Test basis and quadrature
        self.assertEqual(element_data.basis, [basis, basis])
        self.assertEqual(element_data.quadrature, [quad, quad])
        self.assertEqual(element_data.element_name, "grid_name")
        self.assertEqual(element_data_2.basis, [basis, basis])
        self.assertEqual(element_data_2.quadrature, [quad, quad])
        self.assertEqual(element_data_2.element_name, "[B0,(L1I0,L0I0)]")


if __name__ == '__main__':
    unittest.main(verbosity=2)

File: tests/Unit/IO/H5/Test_TensorData.cpp
// Distributed under the MIT License.
// See LICENSE.txt for details.

#include "Framework/TestingFramework.hpp"

#include <cstddef>
#include <string>
#include <vector>

#include "DataStructures/DataVector.hpp"
#include "Domain/Structure/ElementId.hpp"
#include "Framework/TestHelpers.hpp"
#include "IO/H5/TensorData.hpp"
#include "NumericalAlgorithms/Spectral/Basis.hpp"
#include "NumericalAlgorithms/Spectral/Mesh.hpp"
#include "NumericalAlgorithms/Spectral/Quadrature.hpp"
#include "Utilities/GetOutput.hpp"
#include "Utilities/StdHelpers.hpp"

namespace {
template <typename DataType>
void test() {
  TensorComponent tc0("T_x", DataType{8.9, 7.6, -3.4, 9.0});
  TensorComponent tc1("T_y", DataType{8.9, 7.6, -3.4, 9.0});
  TensorComponent tc2("T_x", DataType{8.9, 7.6, -3.4, 9.1});
  CHECK(get_output(tc0) == "(T_x, (8.9,7.6,-3.4,9))");
  CHECK(tc0 == tc0);
  CHECK(tc0 != tc1);
  CHECK(tc0 != tc2);
  CHECK(tc1 != tc2);
  test_serialization(tc0);

  const ElementVolumeData evd0(
      "Element0", {tc0, tc1, tc2}, {2, 2},
      {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
      {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto});
  const auto after_evd0 = serialize_and_deserialize(evd0);
  CHECK(after_evd0.extents == evd0.extents);
  CHECK(after_evd0.tensor_components == evd0.tensor_components);
  CHECK(after_evd0.basis == evd0.basis);
  CHECK(after_evd0.quadrature == evd0.quadrature);
  CHECK(after_evd0.element_name == evd0.element_name);
  CHECK(after_evd0 == evd0);
  // Check operator==
  CHECK(evd0 !=
        ElementVolumeData(
            "Element0", {tc0, tc1, tc2}, {3, 2},
            {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
            {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto}));
  CHECK(evd0 !=
        ElementVolumeData(
            "Element0", {tc2, tc1, tc0}, {2, 2},
            {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
            {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto}));
  CHECK(evd0 !=
        ElementVolumeData(
            "Element0", {tc0, tc1, tc2}, {2, 2},
            {Spectral::Basis::Chebyshev, Spectral::Basis::Chebyshev},
            {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto}));
  CHECK(evd0 != ElementVolumeData(
                    "Element0", {tc0, tc1, tc2}, {2, 2},
                    {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
                    {Spectral::Quadrature::GaussLobatto,
                     Spectral::Quadrature::GaussLobatto}));
  CHECK(evd0 !=
        ElementVolumeData(
            "Element1", {tc0, tc1, tc2}, {2, 2},
            {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
            {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto}));
  CHECK(evd0 ==
        ElementVolumeData(
            "Element0", {tc0, tc1, tc2}, {2, 2},
            {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
            {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto}));
  CHECK(ElementVolumeData(ElementId<1>{0, {{{1, 1}}}}, {tc0, tc1, tc2},
                          Mesh<1>{3, Spectral::Basis::Legendre,
                                  Spectral::Quadrature::GaussLobatto}) ==
        ElementVolumeData("[B0,(L1I1)]", {tc0, tc1, tc2}, {3},
                          {Spectral::Basis::Legendre},
                          {Spectral::Quadrature::GaussLobatto}));
  CHECK(ElementVolumeData(
            ElementId<2>{0, {{{1, 1}, {2, 0}}}}, {tc0, tc1, tc2},
            Mesh<2>{{{3, 4}},
                    {{Spectral::Basis::Legendre, Spectral::Basis::Chebyshev}},
                    {{Spectral::Quadrature::Gauss,
                      Spectral::Quadrature::GaussLobatto}}}) ==
        ElementVolumeData(
            "[B0,(L1I1,L2I0)]", {tc0, tc1, tc2}, {3, 4},
            {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
            {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto}));
  CHECK(ElementVolumeData(
            ElementId<3>{0, {{{1, 1}, {2, 0}, {2, 1}}}}, {tc0, tc1, tc2},
            Mesh<3>{{{3, 4, 5}},
                    {{Spectral::Basis::Legendre, Spectral::Basis::Chebyshev,
                      Spectral::Basis::Legendre}},
                    {{Spectral::Quadrature::Gauss,
                      Spectral::Quadrature::GaussLobatto,
                      Spectral::Quadrature::Gauss}}}) ==
        ElementVolumeData(
            "[B0,(L1I1,L2I0,L2I1)]", {tc0, tc1, tc2}, {3, 4, 5},
            {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev,
             Spectral::Basis::Legendre},
            {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto,
             Spectral::Quadrature::Gauss}));
}
}  // namespace

SPECTRE_TEST_CASE("Unit.DataStructures.TensorData", "[Unit]") {
  test<DataVector>();
  test<std::vector<float>>();
}
